src/partial_quicksync.rs

```diff
@@ -0,0 +1,265 @@
+
+    if previous_hash == &point.hash[..4] {
+      let source_db_path_zst = "backup_source.db.zst";
+      let source_db_path = "backup_source.db";
+
+      if download_file(
+        &client,
+        user_version,
+        point.from,
+        point.to,
+        &point.hash,
+        source_db_path_zst,
+      )
+      .is_err()
+      {
+        download_file(
+          &client,
+          user_version,
+          point.from,
+          point.to,
+          &point.hash,
+          source_db_path,
+        )?;
+      } else {
+        decompress_file(source_db_path_zst, source_db_path)?;
+        fs::remove_file(source_db_path_zst)?;
+      }
+
+      println!("Restoring from {} to {}...", point.from, point.to);
+      let start = Instant::now();
+      execute_restore(&conn, &restore_string)?;
+      fs::remove_file(source_db_path)?;
+      let duration = start.elapsed();
+      println!("Restored {} to {} in {:?}", point.from, point.to, duration);
+    }
+  }
+  Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use anyhow::Result;
+  use rusqlite::Connection;
+  use tempfile::tempdir;
+
+  fn create_test_db() -> Result<(tempfile::TempDir, String)> {
+    let dir = tempdir()?;
+    let db_path = dir.path().join("test.db");
+    let conn = Connection::open(&db_path)?;
+    conn.execute(
+      "CREATE TABLE layers (id INTEGER, applied_block INTEGER, aggregated_hash BLOB)",
+      [],
+    )?;
+    conn.execute(
+      "INSERT INTO layers (id, applied_block, aggregated_hash) VALUES (?, ?, ?), (?, ?, ?)",
+      rusqlite::params![1, 100, 0xAAAA, 2, 200, 0xBBBB],
+    )?;
+    Ok((dir, db_path.to_str().unwrap().to_string()))
+  }
+
+  #[test]
+  fn test_get_base_url() {
+    std::env::set_var("QS_BASE_URL", "https://test.com");
+    assert_eq!(get_base_url(), "https://test.com");
+    std::env::remove_var("QS_BASE_URL");
+    assert_eq!(get_base_url(), DEFAULT_BASE_URL);
+  }
+
+  #[test]
+  fn test_find_start_points() {
+    let metadata = "0,100,aaaa\n101,200,bbbb\n201,300,ijkl\n";
+    let result = find_start_points(150, metadata, 0);
+    assert_eq!(result.len(), 2);
+    assert_eq!(result[0].from, 101);
+    assert_eq!(result[0].to, 200);
+    assert_eq!(result[0].hash, "bbbb");
+
+    let result = find_start_points(150, metadata, 1);
+    println!("{:?}", result);
+    assert_eq!(result.len(), 3);
+    assert_eq!(result[0].from, 0);
+    assert_eq!(result[0].to, 100);
+    assert_eq!(result[0].hash, "aaaa");
+  }
+
+  #[test]
+  fn test_get_latest_from_db() -> Result<()> {
+    let (_dir, db_path) = create_test_db()?;
+    let result = get_latest_from_db(&db_path)?;
+    assert_eq!(result, 2);
+    Ok(())
+  }
+
+  #[test]
+  fn test_get_user_version() -> Result<()> {
+    let (_dir, db_path) = create_test_db()?;
+    let conn = Connection::open(&db_path)?;
+    conn.execute("PRAGMA user_version = 42", [])?;
+    let result = get_user_version(&db_path)?;
+    assert_eq!(result, 42);
+    Ok(())
+  }
```

Comment:
There's no need to return a `Result` from a test.
```suggestion
  #[test]
  fn test_get_latest_from_db() {
    let (_dir, db_path) = create_test_db()?;
    let result = get_latest_from_db(&db_path)?;
    assert_eq!(result, 2);
  }

  #[test]
  fn test_get_user_version() {
    let (_dir, db_path) = create_test_db()?;
    let conn = Connection::open(&db_path)?;
    conn.execute("PRAGMA user_version = 42", [])?;
    let result = get_user_version(&db_path)?;
    assert_eq!(result, 42);
  }
```